cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java

```diff
@@ -123,12 +123,25 @@ public void validateStage(HttpServiceRequest request, HttpServiceResponder respo
     DefaultPipelineConfigurer pipelineConfigurer =
       new DefaultPipelineConfigurer(validatingConfigurer, stageConfig.getName(), Engine.SPARK, stageConfigurer);
 
+    Map<String, String> arguments = Collections.emptyMap();
+
+    // Fetch preferences for this instance and namespace and use them as program arguments if the user selects
+    // this option.
+    if(validationRequest.getResolveMacrosFromPreferences()) {
+      arguments = getContext().getPreferencesForNamespace(namespace, true);
+
+      if (arguments == null) {
+        responder.sendError(HttpURLConnection.HTTP_BAD_REQUEST,
```

Comment:
404 fits this better, 400 is usually for requests that will never succeed, like if the request body is malformed.